### PR TITLE
added model for technical challenge

### PR DESCRIPTION
**Pull request by SantiHi**
## Description

This PR adds the basic model to be iterated on further in the technical challenge. It uses a LSTM (Long Short-Term Memory) model (type of Recurrent Neural Network (RNN)) specifically because it takes into account previous features that could have happened a long time ago, and have gated memory cells allowing the model to learn which features are important or not. 

## Milestones
Technical Challenge 2

## Resources

I hadn't used LSTM's before, so I had a ton of resources that I used to help me:

https://www.youtube.com/watch?v=b61DPVFX03I&t=130s 
https://www.youtube.com/watch?v=YCzL96nL7j0 

- These videos explained what an LSTM was and convinced me to end up using it.

https://codelabs.developers.google.com/codelabs/tfjs-training-regression/index.html#8  - example of tensorflow-js being used but without LSTM specifically, used this for the base model fitting code then modified for LSTM use. 

https://www.tensorflow.org/js/tutorials - tutorials for using the tensorflow js library, including a demo page where I could see other lstm models. 

## Test Plan / demo

This was mostly backend things, but It does now work on the frontend, though the amount the user can specify is still not significant. also, I will be adding in volatility score and a few other aspects to further iterate on this TC. 

Picture of predictions graph: 

image

### CODE REVIEW

File: backend/routes/portfolios.js [hunk 1]
```diff
@@ -342,12 +341,196 @@ router.get("/portfolio/prediction/:id", async (req, res) => {
...
+            ],
+          },
+        });
+      }
     }
   );
 });
 
+/* create and run model! */
+router.post("/model/:id", async (req, res) => {
```
Comment (jomart-gsu) [inline backend/routes/portfolios.js:377]: This is a huge function! For readability it might make sense to split it into even more helpers, like fetch/clean input data, model training, and predicting.

File: backend/routes/portfolios.js [hunk 2]
```diff
@@ -342,12 +341,196 @@ router.get("/portfolio/prediction/:id", async (req, res) => {
...
+    return {
+      date: formatDate(currentDate),
+      price: parseFloat(lastPrice) + offset,
+    };
+  });
+  tf.dispose([X_values, Y_values, input, prediction]);
+  res.json(valuePredict);
+});
+
+// normalize data using xi - min(x) / (max(x) - min(x)) to get data with mean=0 and std 1,
```
Comment (jomart-gsu) [inline backend/routes/portfolios.js:511]: nit: you're normalizing arrays of open/high/low values against close prices, which means the mean won't exactly be zero in some cases. I doubt this will affect model performance meaningfully, but it might be worth thinking through some of the subtler cases here.

File: backend/routes/portfolios.js [hunk 3]
```diff
@@ -342,12 +341,196 @@ router.get("/portfolio/prediction/:id", async (req, res) => {
...
+            ticker: nextEarnings.symbol,
+          },
+          data: {
+            UpcomingEarnings: [
+              nextEarnings.date,
+              String(nextEarnings.epsActual),
+              String(nextEarnings.epsEstimate),
+              String(nextEarnings.revenueActual),
+              String(nextEarnings.revenueEstimate),
+              nextEarnings.symbol,
```
Comment (withlovee) [inline backend/routes/portfolios.js:367]: could this be a json/object instead of array? it is hard to keep track which element represents what.

File: frontend/src/components/predictiontools/LineChart.jsx [hunk 4]
```diff
@@ -22,110 +24,28 @@ ChartJS.register(
...
-  { date: "2025-04-05", price: 125.87 },
-  { date: "2025-04-06", price: 123.93 },
-  { date: "2025-04-07", price: 124.21 },
-  { date: "2025-04-08", price: 125.57 },
-  { date: "2025-04-09", price: 124.41 },
-  { date: "2025-04-10", price: 125.66 },
-];
 
 const LineChart = ({ portfolioData }) => {
+  const [realData, setRealData] = useState(null);
```
Comment (withlovee) [inline frontend/src/components/predictiontools/LineChart.jsx:29]: will we ever have non-realData? maybe name it just `data` or `chartData`?

**Review by withlovee — COMMENTED** [review]
looks good
